### PR TITLE
HAProxy based load balancer in Avalon

### PR DESCRIPTION
**Pull request by pankajgoyal2**
Signed-off-by: pankajgoyal2 <pankaj.goyal@intel.com>

### CODE REVIEW

File: docker/lb/README.md [hunk 1]
```diff
@@ -24,5 +24,34 @@ coming from clients to one of the many avalon-listeners running in backend.
...
+coming from clients to one of the many avalon-listeners running in backend.
+
+## Running
+
+1. Build and run Avalon components.
+    ```bash
+    docker-compose -f docker/lb/docker-compose.yaml -f docker/lb/docker-compose-lb-haproxy.yaml up --build
+
+    The above command is going to start HAProxy based service (avalon-lb-haproxy) along
+    with other avalon services. The HAProxy load balancer configuration is
```
Comment (danintel) [inline docker/lb/README.md:43]: Change avalon to Avalon

Comment (pankajgoyal2) [reply docker/lb/README.md:43]: Ok

**Comment by pankajgoyal2** [comment]
> Avalon already uses ngix for a load balancer.
> I am not sure what this PR does--what is changing in the existing ngix load balancer?

This PR primarily addresses to support HAProxy load balancer. As part of this PR, there is some restructring the way we run nginx based LB .

**Review by TomBarnes — APPROVED** [review]
License headers appear to be in order

File: docker/lb/docker-compose.yaml [hunk 2]
```diff
@@ -175,27 +175,3 @@ services:
...
-      - http_proxy
-      - https_proxy
-      - no_proxy
-    command: |
-      bash -c "
-        service nginx start
-        tail -f /dev/null
-        "
-    stop_signal: SIGKILL
-    depends_on:
```
Comment (manojgop) [inline docker/lb/docker-compose.yaml:198]: We need only depends_on with "avalon-listener-1" for difefret services in this file. Rest we can directly use from top-level docker-compose file ?

File: docker/lb/Dockerfile.haproxy [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+# ------------------------------------------------------------------------------
+
+FROM ubuntu:bionic
+
+RUN apt update \
+ && apt install -y -q \
+    haproxy
+
+COPY haproxy.cfg .
+RUN echo "$(cat haproxy.cfg)" >> /etc/haproxy/haproxy.cfg
```
Comment (manojgop) [inline docker/lb/Dockerfile.haproxy:23]: Can't we directly execute `COPY haproxy.cfg /etc/haproxy/haproxy.cfg`

**Review by manojgop — CHANGES_REQUESTED** [review]

File: docker/lb/Dockerfile.haproxy (hunk 3 above)
Comment (pankajgoyal2) [reply docker/lb/Dockerfile.haproxy:23]: /etc/haproxy/haproxy.cfg has some contents, which I am not changing. I am just appending some extra setting for our purpose.

File: docker/lb/docker-compose.yaml (hunk 2 above)
Comment (pankajgoyal2) [reply docker/lb/docker-compose.yaml:198]: I want to have 3 listener services running. So, do you mean to say that avalon-listener service will come from top level compose file and this file shall have only more listener services like avalon-listener-1/2/3?

File: docker/lb/Dockerfile.haproxy (hunk 3 above)
Comment (manojgop) [reply docker/lb/Dockerfile.haproxy:23]: ok. Please add a comment "Append front and back end load balancer configurations to existing file in /etc/haproxy/haproxy.cfg"

File: docker/lb/docker-compose.yaml (hunk 2 above)
Comment (manojgop) [reply docker/lb/docker-compose.yaml:198]: Yes. That's one possibility to reuse the top level docker-compose.yaml file. use avalon-listerner from top-level compose file and define avaon-listener-1 and avaon-listener-2 in docker-compose-lb.yaml

**Review by Ram-srini — APPROVED** [review]

File: docker/lb/docker-compose-listeners.yaml [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+version: '3.5'
+
+services:
+  avalon-listener-1:
+    container_name: avalon-listener-1
+    image: avalon-listener-dev
+    build:
```
Comment (manojgop) [inline docker/lb/docker-compose-listeners.yaml:21]: we could add dependency of  "avalon-listener" in "avalon-listener-1,2,3". This way we can ensure "avalon-listener-dev" image will be available before starting containers for avalon-listener-1,2,3. We can also remove the entire build section in docker-compose-listener file.

File: docker/lb/Dockerfile.haproxy (hunk 3 above)
Comment (pankajgoyal2) [reply docker/lb/Dockerfile.haproxy:26]: Done

File: docker/lb/docker-compose-listeners.yaml (hunk 4 above)
Comment (pankajgoyal2) [reply docker/lb/docker-compose-listeners.yaml]: Done

File: docker/lb/docker-compose.yaml (hunk 2 above)
Comment (pankajgoyal2) [reply docker/lb/docker-compose.yaml]: Done

**Review by manojgop — APPROVED** [review]

**Pull request merged**